token-handler/src/mint_tokens.rs

```diff
@@ -1,79 +1,187 @@
-use multiversx_sc::api::ESDT_NFT_CREATE_FUNC_NAME;
-use multiversx_sc::types::ManagedArgBuffer;
-use multiversx_sc::types::{ManagedVec, MultiValueEncoded};
+use multiversx_sc::api::{ESDT_MULTI_TRANSFER_FUNC_NAME, ESDT_NFT_CREATE_FUNC_NAME};
+use multiversx_sc::types::{
+    system_proxy, EsdtTokenPayment, ManagedArgBuffer, MultiValueEncoded, ToSelf,
+};
+use multiversx_sc::types::{ManagedVec, TokenIdentifier};
 use multiversx_sc::{codec, err_msg};
-use transaction::OperationEsdtPayment;
+use transaction::{GasLimit, OperationEsdtPayment, StolenFromFrameworkEsdtTokenData, TransferData};
 
-use crate::common;
+const TRANSACTION_GAS: GasLimit = 30_000_000;
+
+use crate::{burn_tokens, common};
 
 #[multiversx_sc::module]
-pub trait MintTokens: utils::UtilsModule + common::storage::CommonStorage {
-    #[endpoint(mintTokens)]
-    fn mint_tokens(&self, operation_tokens: MultiValueEncoded<OperationEsdtPayment<Self::Api>>) {
+pub trait TransferTokensModule:
+    utils::UtilsModule
+    + common::storage::CommonStorage
+    + burn_tokens::BurnTokensModule
+    + tx_batch_module::TxBatchModule
+{
+    // NOTE: will use operation.data.op_sender as well when TransferAndExecuteByUser is implemented
+    #[payable("*")]
+    #[endpoint(transferTokens)]
+    fn transfer_tokens(
+        &self,
+        opt_transfer_data: Option<TransferData<Self::Api>>,
+        to: ManagedAddress,
+        // original_sender: ManagedAddress,
+        tokens: MultiValueEncoded<OperationEsdtPayment<Self::Api>>,
+    ) {
+        let mut output_payments = self.mint_tokens(&tokens.to_vec());
```

Comment:
Let's make `mint_tokens` function to return array of `ManagedVec<EsdtTokenPayment>`, then extend with `call_value_esdt_transfer`. And use this in `distribute_payments`.